realm/realm-library/src/main/cpp/io_realm_SyncSession.cpp

```diff
@@ -49,3 +50,64 @@ JNIEXPORT jboolean JNICALL Java_io_realm_SyncSession_nativeRefreshAccessToken(JN
     CATCH_STD()
     return JNI_FALSE;
 }
+
+
+JNIEXPORT jlong JNICALL Java_io_realm_SyncSession_nativeAddProgressListener(JNIEnv* env, jclass,
+                                                                            jstring j_local_realm_path, jlong listener_id,
+                                                                            jint direction, jboolean is_streaming)
+{
+    try {
+        // JNIEnv is thread confined, so we need a deep copy in order to capture the string in the lambda
+        std::string local_realm_path(JStringAccessor(env, j_local_realm_path));
+        std::shared_ptr<SyncSession> session = SyncManager::shared().get_existing_active_session(local_realm_path);
+        if (!session) {
+            // FIXME: We should lift this restriction
+            ThrowException(env, IllegalState, "Cannot register a progress listener before a session is "
+                                              "created. A session will be created after the first call to Realm.getInstance().");
+            return static_cast<jlong>(0);
+        }
+
+        SyncSession::NotifierType type =
+            (direction == 1) ? SyncSession::NotifierType::download : SyncSession::NotifierType::upload;
+
+        std::function<SyncProgressNotifierCallback> callback = [local_realm_path, listener_id](
+            uint64_t transferred, uint64_t transferrable) {
+            JNIEnv* env = jni_util::JniUtils::get_env(true);
+
+            auto path = env->NewStringUTF(local_realm_path.c_str());
```

Comment:
we are actually compile with c++14, which means we can use move constructor in capture:
```
JavaGlobalRef path_ref;

// JavaGlobalRef's copy constructor should just work if implemented it (which i didn't :P)
std::function<SyncProgressNotifierCallback> callback = [listener_id, path_ref{move(path_ref}] {
    // ....
}
```